File: Samples/Verbum/how-to-create-a-custom-exporter-for-a-specific-format-e3453-16.1.9-/VB/BBCExporter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Text
Imports DevExpress.Utils
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.Export
Imports DevExpress.XtraRichEdit.Internal
Imports DevExpress.XtraRichEdit.Model
Imports System.Drawing
Imports DevExpress.Office
Imports DevExpress.Office.Export
Imports DevExpress.Office.Internal
Imports DevExpress.XtraRichEdit.API.Native.Implementation

Namespace RichEditBBCExporter
	Public Class BBCodeDocumentExporter
		Implements IExporter(Of DocumentFormat, Boolean)
		Friend Shared ReadOnly filter_Renamed As New FileDialogFilter("Bulletin Board Code", "bbc")

        Public ReadOnly Property Filter() As FileDialogFilter Implements IExporter(Of DevExpress.XtraRichEdit.DocumentFormat, Boolean).Filter
            Get
                Return filter_Renamed
            End Get
        End Property
        Public ReadOnly Property Format() As DocumentFormat Implements IExporter(Of DevExpress.XtraRichEdit.DocumentFormat, Boolean).Format
            Get
                Return BBCodeDocumentFormat.Id
            End Get
        End Property
        Public Function SetupSaving() As IExporterOptions Implements IExporter(Of DevExpress.XtraRichEdit.DocumentFormat, Boolean).SetupSaving
            Return New BBCodeDocumentExporterOptions()
        End Function

        Public Function SaveDocument(ByVal documentModel As IDocumentModel, ByVal stream As Stream, ByVal options As IExporterOptions) As Boolean Implements IExporter(Of DevExpress.XtraRichEdit.DocumentFormat, Boolean).SaveDocument
            Dim exporter As New BBCodeExporter(CType(documentModel, DocumentModel), CType(options, BBCodeDocumentExporterOptions))
            exporter.Export(stream)
            Return True
        End Function

		Public Shared Sub Register(ByVal provider As IServiceProvider)
			If provider Is Nothing Then
				Return
			End If
			Dim service As IDocumentExportManagerService = TryCast(provider.GetService(GetType(IDocumentExportManagerService)), IDocumentExportManagerService)
			If service IsNot Nothing Then
				service.RegisterExporter(New BBCodeDocumentExporter())
			End If
		End Sub
	End Class

	Public Class BBCodeDocumentExporterOptions
		Inherits DocumentExporterOptions
		Protected Overrides ReadOnly Property Format() As DocumentFormat
			Get
				Return BBCodeDocumentFormat.Id
			End Get
		End Property
	End Class

	Public Class BBCodeExporter
		Inherits DocumentModelExporter
		Private ReadOnly options_Renamed As BBCodeDocumentExporterOptions
		Private outputStream As Stream
		Private documentContentWriter_Renamed As StreamWriter
		Private hyperlinkExporting As Boolean

		Public Sub New(ByVal documentModel As DocumentModel, ByVal options As BBCodeDocumentExporterOptions)
			MyBase.New(documentModel)
			Guard.ArgumentNotNull(options, "options")
			Me.options_Renamed = options
		End Sub

		Protected ReadOnly Property Options() As BBCodeDocumentExporterOptions
			Get
				Return options_Renamed
			End Get
		End Property
		Protected Friend Property DocumentContentWriter() As StreamWriter
			Get
				Return documentContentWriter_Renamed
			End Get
			Set(ByVal value As StreamWriter)
				documentContentWriter_Renamed = value
			End Set
		End Property

        Public Overridable Shadows Sub Export(ByVal outputStream As Stream)
            Me.outputStream = outputStream
            Using streamWriter As New StreamWriter(outputStream, Encoding.UTF8)
                DocumentContentWriter = streamWriter
                MyBase.Export()
                streamWriter.Flush()
            End Using
        End Sub

        Protected Overrides Sub ExportTextRun(ByVal run As TextRun)
            Dim text As String = run.GetPlainText(PieceTable.TextBuffer)

            If (Not hyperlinkExporting) Then
                If run.FontBold Then
                    DocumentContentWriter.Write("[b]")
                End If
                If run.FontItalic Then
                    DocumentContentWriter.Write("[i]")
                End If
                If run.FontUnderlineType <> UnderlineType.None Then
                    DocumentContentWriter.Write("[u]")
                End If
                If run.FontStrikeoutType <> StrikeoutType.None Then
                    DocumentContentWriter.Write("[s]")
                End If
                If run.ForeColorIndex <> DevExpress.Office.Model.ColorModelInfoCache.EmptyColorIndex Then
                    DocumentContentWriter.Write(String.Format("[color=#{0}]", ColorTranslator.ToHtml(DocumentModel.GetColor(run.ForeColorIndex))))
                End If
                If run.DoubleFontSize <> DocumentModel.DefaultCharacterProperties.DoubleFontSize Then
                    DocumentContentWriter.Write(String.Format("[size={0}]", Math.Min(run.DoubleFontSize, 39)))
                End If
            End If

            DocumentContentWriter.Write(text)

            If (Not hyperlinkExporting) Then
                If run.DoubleFontSize <> DocumentModel.DefaultCharacterProperties.DoubleFontSize Then
                    DocumentContentWriter.Write("[/size]")
                End If
                If run.ForeColorIndex <> DevExpress.Office.Model.ColorModelInfoCache.EmptyColorIndex Then
                    DocumentContentWriter.Write("[/color]")
                End If
                If run.FontStrikeoutType <> StrikeoutType.None Then
                    DocumentContentWriter.Write("[/s]")
                End If
                If run.FontUnderlineType <> UnderlineType.None Then
                    DocumentContentWriter.Write("[/u]")
                End If
                If run.FontBold Then
                    DocumentContentWriter.Write("[/b]")
                End If
                If run.FontItalic Then
                    DocumentContentWriter.Write("[/i]")
                End If
            End If

            MyBase.ExportTextRun(run)
        End Sub

        Protected Overrides Sub ExportFieldCodeStartRun(ByVal run As FieldCodeStartRun)
            MyBase.ExportFieldCodeStartRun(run)
            Dim hyperlinkInfo As HyperlinkInfo = GetHyperlinkInfo(run)
            If hyperlinkInfo Is Nothing Then
                Return
            End If

            DocumentContentWriter.Write(String.Format("[url={0}]", hyperlinkInfo.NavigateUri))
            hyperlinkExporting = True
        End Sub

        Protected Overrides Sub ExportFieldResultEndRun(ByVal run As FieldResultEndRun)
            MyBase.ExportFieldResultEndRun(run)
            If GetHyperlinkInfo(run) Is Nothing Then
                Return
            End If

            DocumentContentWriter.Write("[/url]")
            hyperlinkExporting = False
        End Sub

        Protected Overrides Sub ExportInlinePictureRun(ByVal run As InlinePictureRun)
            DocumentContentWriter.Write(String.Format("[img]{0}[/img]", run.Image.Uri))
            MyBase.ExportInlinePictureRun(run)
        End Sub

        Protected Overrides Function ExportParagraph(ByVal paragraph As Paragraph) As ParagraphIndex
            Dim pi As ParagraphIndex = MyBase.ExportParagraph(paragraph)
            DocumentContentWriter.Write(Environment.NewLine)
            Return pi
        End Function

        Private Function GetHyperlinkInfo(ByVal run As TextRun) As HyperlinkInfo
            Dim runIndex As RunIndex = run.GetRunIndex()
            Dim field As Field = PieceTable.FindFieldByRunIndex(runIndex)
            System.Diagnostics.Debug.Assert(field IsNot Nothing)
            Dim hyperlinkInfo As HyperlinkInfo = Nothing
            If PieceTable.HyperlinkInfos.TryGetHyperlinkInfo(field.Index, hyperlinkInfo) Then
                Return hyperlinkInfo
            End If
            Return Nothing
        End Function
    End Class

	Public NotInheritable Class BBCodeDocumentFormat
		Public Shared ReadOnly Id As New DocumentFormat(431)
	End Class
End Namespace

File: Samples/Verbum/how-to-create-a-custom-exporter-for-a-specific-format-e3453-16.1.9-/CS/BBCExporter.cs
using System;
using System.IO;
using System.Text;
using DevExpress.Utils;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.Export;
using DevExpress.XtraRichEdit.Internal;
using DevExpress.XtraRichEdit.Model;
using System.Drawing;
using DevExpress.Office;
using DevExpress.Office.Export;
using DevExpress.Office.Internal;
using DevExpress.XtraRichEdit.API.Native.Implementation;

namespace RichEditBBCExporter {
    public class BBCodeDocumentExporter : IExporter<DocumentFormat, bool> {
        internal static readonly FileDialogFilter filter = new FileDialogFilter("Bulletin Board Code", "bbc");

        public FileDialogFilter Filter { get { return filter; } }
        public DocumentFormat Format { get { return BBCodeDocumentFormat.Id; } }
        public IExporterOptions SetupSaving() {
            return new BBCodeDocumentExporterOptions();
        }

        public bool SaveDocument(IDocumentModel documentModel, Stream stream, IExporterOptions options) {
            BBCodeExporter exporter = new BBCodeExporter((DocumentModel)documentModel, (BBCodeDocumentExporterOptions)options);
            exporter.Export(stream);
            return true;

        }

        public static void Register(IServiceProvider provider) {
            if(provider == null)
                return;
            IDocumentExportManagerService service = provider.GetService(typeof(IDocumentExportManagerService)) as IDocumentExportManagerService;
            if(service != null)
                service.RegisterExporter(new BBCodeDocumentExporter());
        }
    }

    public class BBCodeDocumentExporterOptions : DocumentExporterOptions {
        protected override DocumentFormat Format { get { return BBCodeDocumentFormat.Id; } }
    }

    public class BBCodeExporter : DocumentModelExporter {
        readonly BBCodeDocumentExporterOptions options;
        Stream outputStream;
        StreamWriter documentContentWriter;
        bool hyperlinkExporting;

        public BBCodeExporter(DocumentModel documentModel, BBCodeDocumentExporterOptions options)
            : base(documentModel) {
            Guard.ArgumentNotNull(options, "options");
            this.options = options;
        }

        protected BBCodeDocumentExporterOptions Options { get { return options; } }
        protected internal StreamWriter DocumentContentWriter { get { return documentContentWriter; } set { documentContentWriter = value; } }

        public virtual void Export(Stream outputStream) {
            this.outputStream = outputStream;
            using(StreamWriter streamWriter = new StreamWriter(outputStream, Encoding.UTF8)) {
                DocumentContentWriter = streamWriter;
                Export();
                streamWriter.Flush();
            }
        }
        protected override void ExportTextRun(TextRun run) {
            string text = run.GetPlainText(PieceTable.TextBuffer);

            if(!hyperlinkExporting) {
                if(run.FontBold)
                    DocumentContentWriter.Write("[b]");
                if(run.FontItalic)
                    DocumentContentWriter.Write("[i]");
                if(run.FontUnderlineType != UnderlineType.None)
                    DocumentContentWriter.Write("[u]");
                if(run.FontStrikeoutType != StrikeoutType.None)
                    DocumentContentWriter.Write("[s]");
                if(run.ForeColorIndex != DevExpress.Office.Model.ColorModelInfoCache.EmptyColorIndex)
                    DocumentContentWriter.Write(string.Format("[color=#{0}]", ColorTranslator.ToHtml(DocumentModel.GetColor(run.ForeColorIndex))));
                if(run.DoubleFontSize != DocumentModel.DefaultCharacterProperties.DoubleFontSize)
                    DocumentContentWriter.Write(string.Format("[size={0}]", Math.Min(run.DoubleFontSize, 39)));
            }

            DocumentContentWriter.Write(text);

            if(!hyperlinkExporting) {
                if(run.DoubleFontSize != DocumentModel.DefaultCharacterProperties.DoubleFontSize)
                    DocumentContentWriter.Write("[/size]");
                if(run.ForeColorIndex != DevExpress.Office.Model.ColorModelInfoCache.EmptyColorIndex)
                    DocumentContentWriter.Write("[/color]");
                if(run.FontStrikeoutType != StrikeoutType.None)
                    DocumentContentWriter.Write("[/s]");
                if(run.FontUnderlineType != UnderlineType.None)
                    DocumentContentWriter.Write("[/u]");
                if(run.FontBold)
                    DocumentContentWriter.Write("[/b]");
                if(run.FontItalic)
                    DocumentContentWriter.Write("[/i]");
            }

            base.ExportTextRun(run);
        }

        protected override void ExportFieldCodeStartRun(FieldCodeStartRun run) {
            base.ExportFieldCodeStartRun(run);
            HyperlinkInfo hyperlinkInfo = GetHyperlinkInfo(run);
            if(hyperlinkInfo == null)
                return;

            DocumentContentWriter.Write(string.Format("[url={0}]", hyperlinkInfo.NavigateUri));
            hyperlinkExporting = true;
        }

        protected override void ExportFieldResultEndRun(FieldResultEndRun run) {
            base.ExportFieldResultEndRun(run);
            if(GetHyperlinkInfo(run) == null)
                return;

            DocumentContentWriter.Write("[/url]");
            hyperlinkExporting = false;
        }

        protected override void ExportInlinePictureRun(InlinePictureRun run) {
            DocumentContentWriter.Write(string.Format("[img]{0}[/img]", run.Image.Uri));
            base.ExportInlinePictureRun(run);
        }

        protected override ParagraphIndex ExportParagraph(Paragraph paragraph) {
            ParagraphIndex pi = base.ExportParagraph(paragraph);
            DocumentContentWriter.Write(Environment.NewLine);
            return pi;
        }

        HyperlinkInfo GetHyperlinkInfo(TextRun run) {
            RunIndex runIndex = run.GetRunIndex();
            Field field = PieceTable.FindFieldByRunIndex(runIndex);
            System.Diagnostics.Debug.Assert(field != null);
            HyperlinkInfo hyperlinkInfo = null;
            if(PieceTable.HyperlinkInfos.TryGetHyperlinkInfo(field.Index, out hyperlinkInfo))
                return hyperlinkInfo;
            return null;
        }
    }

    public static class BBCodeDocumentFormat {
        public static readonly DocumentFormat Id = new DocumentFormat(431);
    }
}
